routers/private/hook.go

```diff
@@ -273,7 +273,12 @@ func HookPreReceive(ctx *macaron.Context, opts private.HookOptions) {
 				}
 			}
 
-			// Detect Protected file pattern
+			var (
+				changedProtectedfiles bool
+				protectedFilePath     string
+			)
+
+			// Check Protected file pattern
```

Comment:
With the changes adding `MergeBlockedByProtectedFiles` to `CheckPRReadyToMerge` this check is duplicated on that branch.

Reply:
How about this way? Thanks!

```DIFF
diff --git a/models/branches.go b/models/branches.go
index 4a83a67dc..acf206d60 100644
--- a/models/branches.go
+++ b/models/branches.go
@@ -212,6 +212,10 @@ func (protectBranch *ProtectedBranch) GetProtectedFilePatterns() []glob.Glob {
 
 // MergeBlockedByProtectedFiles returns true if merge is blocked by protected files change
 func (protectBranch *ProtectedBranch) MergeBlockedByProtectedFiles(pr *PullRequest) bool {
+	if pr.ProtectedFilesChecked {
+		return pr.ChangedProtectedFiles
+	}
+
 	glob := protectBranch.GetProtectedFilePatterns()
 	if len(glob) == 0 {
 		return false
diff --git a/models/pull.go b/models/pull.go
index 9f1f48526..38f8bf843 100644
--- a/models/pull.go
+++ b/models/pull.go
@@ -65,6 +65,9 @@ type PullRequest struct {
 	MergedUnix     timeutil.TimeStamp `xorm:"updated INDEX"`
 
 	isHeadRepoLoaded bool `xorm:"-"`
+
+	ChangedProtectedFiles bool `xorm:"-"`
+	ProtectedFilesChecked bool `xorm:"-"`
 }
 
 // MustHeadUserName returns the HeadRepo's username if failed return blank
diff --git a/routers/private/hook.go b/routers/private/hook.go
index 1489e30fd..5d621e43a 100644
--- a/routers/private/hook.go
+++ b/routers/private/hook.go
@@ -345,6 +345,8 @@ func HookPreReceive(ctx *macaron.Context, opts private.HookOptions) {
 				}
 				// Check all status checks and reviews is ok, unless repo admin which can bypass this.
 				if !perm.IsAdmin() {
+					pr.ProtectedFilesChecked = true
+					pr.ChangedProtectedFiles = changedProtectedfiles
 					if err := pull_service.CheckPRReadyToMerge(pr); err != nil {
 						if models.IsErrNotAllowedToMerge(err) {
 							log.Warn("Forbidden: User %d is not allowed push to protected branch %s in %-v and pr #%d is not ready to be merged: %s", opts.UserID, branchName, repo, pr.Index, err.Error())
```

Reply:
So I think it's now possible that the result of `pull_service.CheckPRReadyToMerge(pr)` may actually be incorrect - in that the `changeProtectedFiles` bit of it might be out-of-date. I think we're gonna need to drop that check from CheckPRReadyToMerge or avoid it somehow.